core/jvm/src/main/scala/zio/blocking/Blocking.scala

```diff
@@ -66,7 +66,11 @@ object Blocking extends Serializable {
      * Locks the specified effect to the blocking thread pool.
      */
     def blocking[R1 <: R, E, A](zio: ZIO[R1, E, A]): ZIO[R1, E, A] =
-      blockingExecutor.flatMap(exec => zio.lock(exec))
+      for {
+        executor <- blockingExecutor
+        here     <- ZIO.effectTotal(executor.here)
```

Comment:
Unfortunately, `here` is not implemented in any executor. I suggest we do this inside `FiberContext`. When doing `Lock`, we can check to see if the provided Executor is the same (`eq`) as the one on the stock. If it is, don't do anything. Otherwise, do the normal thing.

That will fix it for all thread pools, as long as the executors are static. Then we need to make sure ZIO isn't created new executors every time a `def` is called or anything, and then we should be good to go.

Reply:
Excellent point. :smiley:

I updated to do this in `FiberContext` as you suggested. I looked at the constructors for `Executor` and `Blocking` and they appear to generate a single executor so object equality should work. I also manually verified that this works correctly by adding a debug statement to `FiberContext`, though unfortunately it is hard to create a regression test for this.